### ISSUE TITLE
Question: is it possible to determine discrete events where audio samples or music tracks start/stop?

### ISSUE THREAD

**Issue by machty** [snes9xgit/snes9x#639]
I've been doing a lot of weird musical experimentation lately where I try and drive programmatically generated music with weird sources of input.

I [tweeted an idea earlier](https://twitter.com/amatchneer/status/1249004375470673921) about hacking into an SNES emulator in the hopes that I could hook into every time an audio sample started (mario jumping, samus firing a gun), or every time a music track started/stopped, the goal being that you could create alternate soundscapes / audio effects / music from someone playing an SNES game.

I've poked around a bit and am slowly getting a sense of things, but I'm wondering if what I'm trying to do is simply impossible? Is there such a hook/command/instruction for "playing a sample", or is that logic/code already compiled away into the ROM? Are there any parts of the codebase you could direct me to that might clue me in?

**Comment by qwertymodo** [comment]
The SP700 is a full-fledged reprogrammable 6502ish CPU. The commands for controlling SFX and BGM are completely custom-written programs, and are different for every game (though many games from the company often reused the same code with only minor changes from one game to the next).  You can look into existing MSU-1 hacks for an idea of what is possible currently. Taking it further to hook SFX as well isn't currently possible, but the work involved would be similar if it was.

**Comment by machty** [comment]
@qwertymodo thank you for your prompt and helpful response :)

**Comment by machty** [comment]
Just wanted to share where I ended up with this idea: https://www.instagram.com/p/B-9mBzWju8g/

Basically, I hooked into "KON" (key-on) events, and sent the pitch to one of my music programs via OSC. It gets pretty tricky though since "pitch" is just the rate at which you move through the waveform, and you don't have any knowledge whether what they're playing is a sound effect or a musical note, and if it's a musical note, you don't know what the original frequency/pitch of the note was originally recorded. Still, ended up being a fun project.
